java/daikon/dcomp/BuildJDK.java

```diff
@@ -3,375 +3,500 @@
+      System.exit(1);
+    }
 
-  void copyStreams(InputStream fis, OutputStream fos) throws Exception {
-    byte[] buf = new byte[1024];
-    int i = 0;
-    while ((i = fis.read(buf)) != -1) {
-      fos.write(buf, 0, i);
+    JAVA_HOME = jrt.getAbsolutePath();
+    if (!java_home.startsWith(JAVA_HOME)) {
+      System.out.printf(
+          "JAVA_HOME (%s) does not agree with java.home (%s).%n", JAVA_HOME, java_home);
+      System.out.printf("Please correct your Java environment.%n");
+      System.exit(1);
     }
-    fis.close();
-    fos.close();
   }
 
-  void translate_classes(JarFile jfile, String dest, String prefix, String prefixOfFilesToInclude)
-      throws java.io.IOException {
-
-    // Map from classname to class so we can find out information about
-    // classes we have not yet instrumented.
-    Map<String, JavaClass> classmap = new LinkedHashMap<>();
+  /**
+   * For Java 8 the Java runtime is located in rt.jar. This method creates an InputStream for each
+   * class in rt.jar and saves this information in {@link #class_stream_map}.
+   */
+  void gather_runtime_from_jar() {
 
+    String jar_name = java_home + "/lib/rt.jar";
+    System.out.printf("using jar file %s%n", jar_name);
     try {
-
-      // Create the destination directory
-      File dfile = new File(dest);
-      dfile.mkdirs();
-
+      JarFile jfile = new JarFile(jar_name);
       // Get each class to be instrumented and store it away
       Enumeration<JarEntry> entries = jfile.entries();
       while (entries.hasMoreElements()) {
         JarEntry entry = entries.nextElement();
         // System.out.printf("processing entry %s%n", entry);
         final String entryName = entry.getName();
-        if (!entryName.startsWith(prefixOfFilesToInclude) && !entryName.startsWith("META-INF")) {
+        if ((entryName.endsWith("/")) || (entryName.endsWith("~"))) {
           continue;
         }
-        if ((entryName.endsWith("/")) || (entryName.endsWith("~"))) {
+
+        // Get the InputStream for this file
+        InputStream is = jfile.getInputStream(entry);
+        class_stream_map.put(entryName, is);
+      }
+    } catch (Exception e) {
+      throw new Error("Problem while reading " + jar_name, e);
+    }
+  }
+
+  /**
+   * For Java 9+ the Java runtime is located in a series of modules. At this time, we are only
+   * pre-instrumenting the java.base module. This method initializes the DirectoryStream used to
+   * explore java.base. It calls gather_runtime_from_modules_directory to process the directory
+   * structure.
+   */
+  void gather_runtime_from_modules() {
+
+    FileSystem fs = FileSystems.getFileSystem(URI.create("jrt:/"));
+    Path modules = fs.getPath("/modules");
+    System.out.printf("using modules file %s%n", java_home + "/lib/modules");
+    try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(modules, "java.base*")) {
+      for (Path module : directoryStream) {
```

Comment:
Since I think this is a directory, I would name it `moduleDir` rather than `module`.